src/Sparrow/Threading/MultipleUseFlag.cs

```diff
@@ -0,0 +1,101 @@
+﻿using System;
+using System.Runtime.CompilerServices;
+using System.Threading;
+
+namespace Sparrow.Threading
+{
+    /// <summary>
+    /// A thread-safe, multiple use flag that can be raised and lowered at will; meant to be
+    /// single-user.
+    /// </summary>
+    /// 
+    /// Example use case is one class has multiple threads that access it and change the
+    /// behavior externally, such as "stop logging".
+    /// 
+    /// For convincing on why you should use this class instead of rolling your own, see
+    /// http://blog.alexrp.com/2014/03/30/dot-net-atomics-and-memory-model-semantics/ and
+    /// http://issues.hibernatingrhinos.com/issue/RavenDB-8260 .
+    /// 
+    /// PERF: This is a struct instead of a class so that its usage may be made invisible. Do
+    /// NOT change this without good reason, could have sizeable impact.
+    public struct MultipleUseFlag
```

Comment:
I'm worried about people passing this around and not realizing that this will copy

Reply:
Added a warning in caps to prevent it.

Another thing I could do is add a VALIDATE check, I'm thinking of something along the lines of:

- Every instance holds a `class FlagWatcher`, which receives `this` as a parameter to the constructor.
- Every access to the internal variable is notified to the `FlagWatcher`, along with the instance that is doing it.
- If the instance is not the one that created it, then it kills the process.

This should help ensure no one is copying it. Do you have any other idea?

Reply:
No one will read the warning :-)
The idea of adding this as `VALIDATE` / `DEBUG` would be better

Reply:
Will implement this on Monday, turned out to be harder than I expected.